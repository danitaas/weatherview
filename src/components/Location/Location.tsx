import * as React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { IWeatherAction, loadWeather, LoadWeatherStart } from "../../logic/weather";
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import Paper from "@material-ui/core/Paper";

export interface IProps {
    onSubmit?: any;
    dispatch?: any; // injected by connect()
    classes?: any;
}

export interface IState {
    location: string;
    countrycode: string;
}

export class Location extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = { location: "London", countrycode: "uk" };

        this.handleLocationChange = this.handleLocationChange.bind(this);
        this.handleCountryChange = this.handleCountryChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    private handleLocationChange(event: any) {
        console.log("handleLocationChange", event.target.value);
        this.setState({ location: event.target.value });
    }

    private handleCountryChange(event: any) {
        console.log("handleCountryChange", event.target.value);
        this.setState({ countrycode: event.target.value });
    }

    private handleSubmit(event: any) {
        console.log("handleSubmit");
        if(event) {
            event.preventDefault();
        }

        //could be actioncreator injected by connect or inline
        // this.props.onSubmit(LoadWeatherStart(this.state.location, this.state.countrycode));
        // this.props.dispatch(loadWeather(this.state.location, this.state.countrycode));
        this.props.onSubmit(this.state.location, this.state.countrycode);
    }

    public render() {
        return (
            <div>
                <form id="searchform" onSubmit={this.handleSubmit}>
                    <Paper className={this.props.classes.container}>
                        <div>
                            <TextField
                                id="location"
                                label="Location"
                                className={this.props.classes.textField}
                                value={this.state.location}
                                onChange={this.handleLocationChange}
                                margin="normal"
                            />
                        </div>
                        <div>
                            <FormControl className={this.props.classes.formControl}>
                                <InputLabel htmlFor="age-countrycode">Country</InputLabel>
                                <Select
                                    value={this.state.countrycode}
                                    onChange={this.handleCountryChange}
                                    inputProps={{
                                        name: 'countrycode',
                                        id: 'countrycode',
                                    }}
                                >
                                    <MenuItem value="">
                                        <em>Choose</em>
                                    </MenuItem>
                                    <MenuItem value={"uk"}>UK</MenuItem>
                                    <MenuItem value={"canada"}>Canada</MenuItem>
                                </Select>
                            </FormControl>
                        </div>
                        <div>
                            <Button type="submit" variant="raised" color="primary">
                                Search
                            </Button>
                        </div>
                    </Paper>
                </form>
            </div>
        );
    }
}

export function mapStateToProps(state: any) {
    return {
        // todo: could bind initial location/country code from redux state
    };
}

export function mapDispatchToProps(dispatch: any) {
    return {
        onSubmit: (location: string, countrycode: string) => {
            //todo: use redux-saga to control async logic
            dispatch(LoadWeatherStart(location, countrycode));
            dispatch(loadWeather(location, countrycode));
        },
        dispatch,
    };
}

export const styles: any = (theme: any) => ({
    container: {
        display: 'flex',
        flexWrap: 'wrap',
        alignItems: 'center',
        maxWidth: 600,
        margin: '0 auto',
        justifyContent: 'space-around',
    },
    formControl: {
        margin: theme.spacing.unit,
        minWidth: 120,
    },
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 200,
    },
    menu: {
        width: 200,
    },
});

// separate variables for hot reloading + test
const component = Location;
const connectedcomponent: any = connect<IProps, IProps, IProps>(
    mapStateToProps,
    mapDispatchToProps
)(component);
const styledcomponent = withStyles(styles)(connectedcomponent);
// todo: add other HOCs here
export default styledcomponent;
