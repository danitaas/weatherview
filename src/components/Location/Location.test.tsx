import * as React from "react";
import { Location } from "./Location";
// todo: cant use typescript import without correcting toolchain
const ReactShallowRenderer = require("react-test-renderer/shallow");
const ReactTestRenderer = require("react-test-renderer");
import * as enzyme from 'enzyme';

it("renders shallow without crashing", () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<Location />);
    const result = renderer.getRenderOutput();
    expect(result.type).toBe("div");
});

it("renders snapshot", () => {
    const tree = ReactTestRenderer.create(<Location />).toJSON();
    expect(tree).toMatchSnapshot();
});

// it("handleSubmit - jest", () => {
//     const mockcall = jest.fn();
//     const tree = ReactTestRenderer.create(<Location onSubmit={mockcall} />).toJSON();
//     wrapper.find('button').at(0).simulate('click');
//     expect(mockLogout).toHaveBeenCalled();
// });

it("handleSubmit - enzyme", () => {
    let clicked = false;
    const handleSubmit = () => {
        console.log("handleSubmit");
        clicked = true;
    };
    const wrapper = enzyme.shallow(<Location onSubmit={handleSubmit} />);
    const button = wrapper.find('#searchform');
    button.simulate('submit');
    expect(clicked).toBe(true);
});

