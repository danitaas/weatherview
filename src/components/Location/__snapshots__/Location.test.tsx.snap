// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
<div>
  <form
    id="searchform"
    onSubmit={[Function]}
  >
    <div
      className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2 display flexWrap alignItems maxWidth margin justifyContent"
    >
      <div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-28 MuiFormControl-marginNormal-29 marginLeft marginRight width"
        >
          <label
            className="MuiFormLabel-root-37 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-36 MuiInputLabel-shrink-35"
            data-shrink={true}
            htmlFor="location"
          >
            Location
          </label>
          <div
            className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-50"
              defaultValue={undefined}
              disabled={false}
              id="location"
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value="London"
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="MuiFormControl-root-28 margin minWidth"
        >
          <label
            className="MuiFormLabel-root-37 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-36 MuiInputLabel-shrink-35"
            data-shrink={true}
            htmlFor="age-countrycode"
          >
            Country
          </label>
          <div
            className="MuiInput-root-42 MuiInput-formControl-43 MuiInput-underline-46"
          >
            <div
              className="MuiSelect-root-55"
            >
              <div
                aria-haspopup="true"
                aria-owns={null}
                aria-pressed="false"
                className="MuiSelect-select-56 MuiSelect-selectMenu-57 MuiInput-input-50"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                UK
              </div>
              <input
                autoComplete={undefined}
                autoFocus={undefined}
                defaultValue={undefined}
                id="countrycode"
                name="countrycode"
                onKeyDown={undefined}
                onKeyUp={undefined}
                placeholder={undefined}
                rows={undefined}
                type="hidden"
                value="uk"
              />
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-60 MuiSelect-icon-59"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          className="MuiButtonBase-root-95 MuiButton-root-71 MuiButton-contained-80 MuiButton-containedPrimary-81 MuiButton-raised-83 MuiButton-raisedPrimary-84"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-72"
          >
            Search
          </span>
          <span
            className="MuiTouchRipple-root-98"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
